wagtail/contrib/modeladmin/tests/test_modeladmin_edit_handlers.py

```diff
@@ -0,0 +1,136 @@
+
+
+class TestExtractPanelDefinitionsFromModelAdmin(TestCase, WagtailTestUtils):
+    """tests that edit_handler and panels can be defined on modeladmin"""
+
+    def setUp(self):
+        self.login()
+
+    def test_model_edit_handler(self):
+        # tests that edit_handler defintion from model is used to create
+        # a model instance if edit_handler is set on the model
+        response = self.client.post('/admin/modeladmintest/person/create/', {
+            'first_name': "John",
+            'last_name': "Doe",
+            'address': "123 Main St Anytown",
+            'phone_number': "+123456789"
+        })
+        # Should redirect back to index
+        self.assertRedirects(response, '/admin/modeladmintest/person/')
+
+        # Check that the person was created
+        self.assertEqual(Person.objects.filter(
+            first_name="John",
+            last_name="Doe",
+            phone_number="+123456789"
+        ).count(), 1)
+
+        # verify that form fields are returned which have been defined
+        # in Person.edit_handler
+        response = self.client.get('/admin/modeladmintest/person/create/')
+        self.assertIn('first_name', response.content.decode('UTF-8'))
+        self.assertIn('last_name', response.content.decode('UTF-8'))
+        self.assertIn('phone_number', response.content.decode('UTF-8'))
+        self.assertNotIn('address', response.content.decode('UTF-8'))
+        self.assertEqual(
+            [ii for ii in response.context['form'].fields],
+            ['first_name', 'last_name', 'phone_number']
+        )
+
+    def test_model_panels(self):
+        # tests that panel definition from model is used to create
+        # a model instance if no edit_handler is set on the model
+        response = self.client.post('/admin/modeladmintest/friend/create/', {
+            'first_name': "John",
+            'last_name': "Doe",
+            'address': "123 Main St Anytown",
+            'phone_number': "+123456789"
+        })
+        # Should redirect back to index
+        self.assertRedirects(response, '/admin/modeladmintest/friend/')
+
+        # Check that the friend was created
+        self.assertEqual(Friend.objects.filter(
+            first_name="John",
+            phone_number="+123456789"
+        ).count(), 1)
+
+        # verify that form fields are returned which have been defined
+        # in Friend.panels
+        response = self.client.get('/admin/modeladmintest/friend/create/')
+        self.assertIn('first_name', response.content.decode('UTF-8'))
+        self.assertNotIn('last_name', response.content.decode('UTF-8'))
+        self.assertIn('phone_number', response.content.decode('UTF-8'))
+        self.assertNotIn('address', response.content.decode('UTF-8'))
+        self.assertEqual(
+            [ii for ii in response.context['form'].fields],
+            ['first_name', 'phone_number']
+        )
+
+    def test_model_admin_edit_handler(self):
```

Comment:
@mr-stateradio Each of these tests seems to be quite clearly split into 2 parts:

1. Post the form and check the object that gets created.
2. Loads the 'create' view and checks what fields are available. 

I would suggest that part 2 is really the most useful part here, and is probably really all that is needed. If you have any experience of using the [mock](https://docs.python.org/3/library/unittest.mock.html) library in Python, then an added bonus would be to show that both `ModelFormView.get_edit_handler()` and `ModelAdmin.get_edit_handler()` are called in order to render the page - But, I certainly wouldn't class that as necessary.

The state of a newly created instance after posting could potentially be affected by a number of things other than what `edit_handler` was used by the view (for example: default field values / custom clean/save logic on the model), and without going and checking those things, it's difficult to know what the expected output should be.

Reply:
@ababic I restricted the test cases to check only for the fields available in the 'create view', and used the mock Library to verify the appropriate `get_edit_handler` methods are called. Can you check the file `test_modeladmin_edit_handlers.TestExtractPanelDefinitionsFromModelAdmin` if we still have some potential for improvement or if we are good to go :)

Reply:
@mr-stateradio This is better, thank you. Although, I'm not sure that we need to confirm that both methods are called for each different URL as a separate test. I think we can assume that the view works the same way, regardless of what the model is. I was just thinking once, at the beginning of the class, to help provide some context to the reader in regards to the other tests. Sorry if I could've been clearer there.

Reply:
@mr-stateradio I could be wrong, but it appears that we're not really testing that values on the ModelAdmin class are preferred over values on the model when both are defined, which I think is probably an important detail of the implementation, and should have a test to help guard against regressions. 

Would you be okay to add something there? I'm guessing it might be easier to take a more  'unit test like' approach (create a `ModelAdmin` instance in the test, change it's attributes, play with the `get_edit_handler()` method directly).

Reply:
@ababic I just posted a newer version of the `test_model_form_view_edit_handler_called` and `test_model_form_view_edit_handler_called` since the version you reviewed was not Python 3.5 compatible and had some other issues. If you check the new version you will see that I check different things there, please check again if you are fine with them now.

You are absolutely right regarding your request for adding a test to verify that values on the ModelAdmin class are preferred over values on the model. Will provide another test case for it. Will try to make this new test case more "unit test like", but will keep the others for now as they are. Maybe we adapt the other ones in the next feedback loop :)

Reply:
@mr-stateradio Thank you, yes. These are much better now.